client/src/controllers/TabsController.ts

```diff
@@ -0,0 +1,319 @@
+/* eslint-disable no-shadow */
+import { Controller } from '@hotwired/stimulus';
+import { debounce } from '../utils/debounce';
+
+interface IndexedEventTarget extends EventTarget {
+  index: number;
+}
+
+interface TabLink extends HTMLAnchorElement {
+  index: number;
+}
+
+/**
+ * @example - creating a simple tabs interface
+ *<div class="w-tabs" data-controller="w-tabs" data-action="popstate@window->w-tabs#loadHistory" data-w-tabs-selected-value="" data-w-tabs-selected-class="animate-in">
+ *    <div class="w-tabs__list" role="tablist" data-w-tabs-target="list">
+ *        <a id="tab-label-tab-1" href="#tab-tab-1" class="w-tabs__tab" role="tab" tabindex="-1"
+ *           data-action="click->w-tabs#handleTabChange:prevent keydown.right->w-tabs#selectNext keydown.left->w-tabs#selectPrevious keydown.home->w-tabs#selectFirst keydown.end->w-tabs#selectLast" data-w-tabs-target="label">
+ *            Tab 1
+ *        </a>
+ *        <a id="tab-label-tab-2" href="#tab-tab-2" class="w-tabs__tab" role="tab"
+ *           data-action="click->w-tabs#handleTabChange:prevent keydown.right->w-tabs#selectNext keydown.left->w-tabs#selectPrevious keydown.home->w-tabs#selectFirst keydown.end->w-tabs#selectLast" data-w-tabs-target="label">
+ *            Tab 2
+ *        </a>
+ *    </div>
+ *    <div class="tab-content tab-content--comments-enabled">
+ *        <section id="tab-tab-1" class="w-tabs__panel " role="tabpanel" aria-labelledby="tab-label-tab-1" data-w-tabs-target="panel">
+ *            tab-1
+ *        </section>
+ *        <section id="tab-tab-2" class="w-tabs__panel " role="tabpanel" aria-labelledby="tab-label-tab-2" data-w-tabs-target="panel">
+ *            tab-2
+ *        </section>
+ *    </div>
+ *</div>
+ */
+
+export class TabsController extends Controller<HTMLDivElement> {
```

Comment:
Let's try to clean this up a bit.

1. Remove classes in the example HTML as this adds noise and is not strictly part of the 'Stimulus contract' for use.
2. Add some of the notes/description from the original code, it's super helpful.
3. Move the keyboard listeners to the tabslist.
4. Clean up the HTML nesting.
5. Ensure there is no line break above the controller, many editors will simply ignore this JSDoc unless its directly above the class/function.


```suggestion
/**
 * Adds the ability for the controlled elements to behave as selectable tabs.
 *
 * All tabs and tab content must be nested in an element within the scope of the controller.
 * All tab buttons need the role="tab" attr and an href with the tab content ID with the target 'label'.
 * Tab contents need to have the role="tabpanel" attribute and and ID attribute that matches the href of the tab link with the target 'panel'.
 * Tab buttons should also be wrapped in an element with the role="tablist" attribute.
 * Use the target 'trigger' on an Anchor link and set the href to the #ID of the tab you would like to trigger.
 *
 * @example
 * ```html
 * <div data-controller="w-tabs" data-action="popstate@window->w-tabs#loadHistory" data-w-tabs-selected-class="animate-in">
 *   <div role="tablist" data-action="click->w-tabs#handleTabChange:prevent keydown.right->w-tabs#selectNext keydown.left->w-tabs#selectPrevious keydown.home->w-tabs#selectFirst keydown.end->w-tabs#selectLast">
 *     <a id="tab-label-tab-1" href="#tab-tab-1" role="tab" data-w-tabs-target="label">
 *       Tab 1
 *     </a>
 *     <a id="tab-label-tab-2" href="#tab-tab-2" role="tab" data-w-tabs-target="label">
 *       Tab 2
 *     </a>
 *    </div>
 *    <div class="tab-content tab-content--comments-enabled">
 *     <section id="tab-tab-1" role="tabpanel" aria-labelledby="tab-label-tab-1" data-w-tabs-target="panel">
 *       Tab 1 content
 *     </section>
 *     <section id="tab-tab-2" role="tabpanel" aria-labelledby="tab-label-tab-2" data-w-tabs-target="panel">
 *       Tab 2 content
 *     </section>
 *    </div>
 *</div>
 */
export class TabsController extends Controller<HTMLDivElement> {
```